### PR TITLE
feat: replace five_gnb_identity with fiveg_core_gnb and extend fiveg_f1

### PR DESCRIPTION
**Pull request by patriciareinoso**
# Description

This PR depends on this library https://github.com/canonical/sdcore-nms-k8s-operator

- Remove  `fiveg_gnb_identity`  library
- Replace `fiveg_gnb_identity` with `fiveg_core_gnb`.
- Observe the relation created and relation changed events for the `fiveg_core_gnb` relation and use the centralized `_configure` method to handle it.
- Publish the `gnb_name` in `fiveg_core_gnb`.
- Take the information from `fiveg_core_gnb` (TAC and PLMNs) and use it to build the config file and to fil the `fiveg_f1` relation.
- Remove `mcc`, `mnc`, `sst`, `tac` from charm config since they are no longer used.
- SD was not set before. This element is added to the config file and casted to hex by the template.
- Update the config file to allow multiple PLMNs

## Integration test
 
- The integration test is simplified to avoid dependecies with the NMS. `any_charm` (A charm simulates any kind of relation consumer charm to make it easy to write integration tests for relation provider charms) is used.
- `any_charm` mocks the NMS so it implments the  `sdcore_config` and `fiveg_core_gnb` in the integration tests.
- The `sdcore_config` and `fiveg_core_gnb` relations were added to `any_charm` [here](https://github.com/canonical/any-charm/pull/31)
- This is not the "prettiest" solution but it reduces maintenance cost and code duplication.

## The logic 
- The CU publishes the `gnb_name` in `fiveg_core_gnb` after its container and k8s requirements are met but before TAC and PLMNs are available.
- Once it gets the `fiveg_core_gnb` information (TAC and PLMNs), it writes the configuration file and publish the F1 information
- The charm stays in `waiting` status until the `fiveg_core_gnb` information (TAC and PLMNs) is available.

image

# Checklist:

- [ ] My code follows the [style guidelines](/CONTRIBUTING.md) of this project
- [ ] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that validate the behaviour of the software
- [ ] I validated that new and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have bumped the version of the library

### CODE REVIEW

File: tests/unit/test_charm_collect_status.py [hunk 1]
```diff
@@ -202,6 +209,57 @@ def test_given_n3_route_is_missing_when_collect_unit_status_then_status_is_waiti
...
+    @pytest.mark.parametrize(
+        "tac,plmns",
+        [
+            pytest.param(None, [PLMNConfig(mcc="001", mnc="01", sst=31)], id="tac_is_none"),
+            pytest.param(23, None, id="plmns_is_none"),
+            pytest.param(None, None, id="plmns_and_tac_are_none"),
+        ],
+    )
+    def test_given_fiveg_core_gnb_tac_and_plmns_are_none_when_collect_unit_status_then_status_is_waiting(  # noqa: E501
+        self, tac, plmns
```
Comment (patriciareinoso) [inline tests/unit/test_charm_collect_status.py:221]: Need to add the fiveg_core_gnb relation.
Stay  blocked if the relation does not exits and waiting if the remote information is not there?

File: src/charm.py [hunk 2]
```diff
@@ -223,6 +228,10 @@ def _relation_created(self, relation_name: str) -> bool:
         return bool(self.model.relations.get(relation_name))
 
     def _generate_cu_config(self) -> str:
+        tac = self._core_gnb_requirer.tac
+        plmns = self._core_gnb_requirer.plmns
+        if not tac or not plmns:
+            return ""
```
Comment (Gmerold) [inline src/charm.py:234]: I guess we should leave a trace in the logs if it fails here.

File: src/charm.py [hunk 3]
```diff
@@ -377,42 +384,35 @@ def _configure_pebble(self, restart=False) -> None:
...
+        if not tac or not plmns:
+            return
+
         self._f1_provider.set_f1_information(
             ip_address=f1_ip.split("/")[0],
             port=self._charm_config.f1_port,
-            tac=HARDCODED_TAC,
-            plmns=HARDCODED_PLMNS,
+            tac=tac,
+            plmns=plmns,  # type: ignore
```
Comment (Gmerold) [inline src/charm.py:415]: Can we not ignore static analysis here?

File: src/templates/cu.conf.j2 [hunk 4]
```diff
@@ -11,12 +11,24 @@ gNBs =
         tracking_area_code  = {{ tac }};
         plmn_list =
         (
+            {% for plmn in plmn_list -%}
             {
-                mcc = {{ mcc }};
-                mnc = {{ mnc }};
+                mcc        = {{ plmn.mcc }};
```
Comment (Gmerold) [inline src/templates/cu.conf.j2:16]: Why all these extra spaces?

File: tests/integration/test_integration.py [hunk 5]
```diff
@@ -17,8 +20,7 @@
 GRAFANA_AGENT_CHARM_NAME = "grafana-agent-k8s"
 NRF_CHARM_NAME = "sdcore-nrf-k8s"
 NRF_CHARM_CHANNEL = "1.5/edge"
```
Comment (Gmerold) [inline tests/integration/test_integration.py:22]: Not strictly related to this PR, but we should probably update to `1.6/edge`.

File: tests/unit/resources/expected_config.conf [hunk 6]
```diff
@@ -12,10 +12,15 @@ gNBs =
         plmn_list =
         (
             {
-                mcc = 001;
-                mnc = 01;
+                mcc        = 001;
```
Comment (Gmerold) [inline tests/unit/resources/expected_config.conf:15]: Same question about the extra spaces.

**Review by Gmerold — COMMENTED** [review]
Other than some minor comment, LGTM

File: tests/integration/test_integration.py (hunk 5 above)
Comment (patriciareinoso) [reply tests/integration/test_integration.py:22]: good catch

**Review by Gmerold — APPROVED** [review]

**Pull request merged**